impeller/entity/geometry/ellipse_geometry.cc

```diff
@@ -0,0 +1,102 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "impeller/entity/geometry/ellipse_geometry.h"
+
+#include "flutter/impeller/entity/geometry/circle_tessellator.h"
+
+namespace impeller {
+
+EllipseGeometry::EllipseGeometry(Point center, Scalar radius)
+    : center_(center), radius_(radius) {
+  FML_DCHECK(radius >= 0);
+}
+
+GeometryResult EllipseGeometry::GetPositionBuffer(
+    const ContentContext& renderer,
+    const Entity& entity,
+    RenderPass& pass) const {
+  auto& host_buffer = pass.GetTransientsBuffer();
+
+  VertexBufferBuilder<SolidFillVertexShader::PerVertexData> vtx_builder;
+
+  CircleTessellator tessellator(entity.GetTransform(), radius_);
+  vtx_builder.Reserve(tessellator.GetCircleVertexCount());
```

Comment:
I would avoid using the vtx builder if you know the exact size. You can skip the intermediate buffer (well, the second of two intermediate buffers) and emplace right into the host buffer with

https://github.com/flutter/engine/blob/main/impeller/core/host_buffer.h#L109

Reply:
HostBuffer doesn't have a reserve method, though.

Reply:
HostBuffer will allocate blocks of memory and grow as needed. There is no magic to emplacing into the host buffer via VertexBufferBuilder, you're doing the exact same allocations twice regardless of the reserve call.

Reply:
Oh, I thought you were saying to place each vertex into the buffer individually, in which case it would want a reserve method.

I still need to allocate space for the tessellator to insert into. Before I was allocating vectors and reserving into them and then there was some ugly glue code to Emplace that vector into the HostBuffer. I switched to using the vertex builders because they cleaned up the code and I didn't notice any performance difference.

Now, if I could use the HostBuffer as my array to insert into from the tessellator, that would be nice, but there is no "individual emplace" method on it and no Reserve method on it.

Reply:
The method I linked requires you to specify the required storage so that the host buffer can ensure it has sufficient space

Reply:
I understand, but it also copies the data in which means you have to have the data ready to go in some kind of storage somewhere. If I get rid of vtx builder, then I have to use a vector to store things, either way I have to reserve space in some kind of storage.

I used to use a vector, but didn't see much performance difference compared to vtx builder and vtx builder was better typed and hid the details of emplacement better.

Reply:
Oh, I was missing the Emplace that takes a callback. Now I see how this could work.

I'm wondering, though, why the callback is needed when you get back a BufferView?

Reply:
So, something like this, which seems to work fine even for UV buffers?

```
GeometryResult EllipseGeometry::GetPositionUVBuffer(
    Rect texture_coverage,
    Matrix effect_transform,
    const ContentContext& renderer,
    const Entity& entity,
    RenderPass& pass) const {
  auto& host_buffer = pass.GetTransientsBuffer();
  using VT = TextureFillVertexShader::PerVertexData;

  auto uv_transform =
      texture_coverage.GetNormalizingTransform() * effect_transform;

  Scalar radius = radius_;
  const Point& center = center_;
  CircleTessellator tessellator(entity.GetTransform(), radius_);
  size_t count = tessellator.GetCircleVertexCount();
  auto vertex_buffer = host_buffer.Emplace(count * sizeof(VT), alignof(VT),
      [&tessellator, &uv_transform, &center, radius](uint8_t* buffer) {
        auto vertices = reinterpret_cast<VT*>(buffer);
        tessellator.GenerateCircleTriangleStrip(
            [&vertices, &uv_transform](const Point& p) {  //
              *vertices++ = {
                  .position = p,
                  .texture_coords = uv_transform * p,
              };
            },
            center, radius);
      });

  return GeometryResult{
      .type = PrimitiveType::kTriangleStrip,
      .vertex_buffer = {
          .vertex_buffer = vertex_buffer,
          .vertex_count = count,
          .index_type = IndexType::kNone,
      },
      .transform = Matrix::MakeOrthographic(pass.GetRenderTargetSize()) *
                   entity.GetTransform(),
      .prevent_overdraw = false,
  };
}
```

Reply:
Done